### PR TITLE
Implement Batch Query Commitements RFC

### PR DESCRIPTION
**Pull request by sergiimk**
## Description

Closes: https://github.com/kamu-data/kamu-node/issues/125
Closes: https://github.com/kamu-data/kamu-node/issues/126

Implements response signing and a new commitment scheme [according to the RFC](https://docs.google.com/document/d/1JIuI5dd6aSYiBZpXZOmv5DGS4eFKNHQ3jSsfSsDxL00/edit).

Note that the old response format is preserved for compatibility via `V1 / V2` enum trick. If user request in the new format (e.g. by including `?include=proof`) only then they will get a response in the new format. This will allow us to transition to the new format progressively, without breaking the oracles for example.

Also introduces a new `{dataset}/metadata` endpoint that allows getting dataset schema, license, info etc.

## Checklist before requesting a review

- [x] Unit and integration tests added
  
- [x] Compatibility:
    
  - [x] Network APIs: ✅
    Extra steps taken to preserve the compatibility
  - [x] Workspace layout and metadata: ✅ 
    New config options added, but they have defaults
  - [x] Configuration: ✅
    
  - [x] Container images: ✅
- [x] Documentation:
    
  - [x] [Changelog](./CHANGELOG.md): ✅
    
  - [x] [Public documentation](https://github.com/kamu-data/kamu-docs/): ❌
    Will be done separately
- [x] Downstream effects:
    
  - [x] [kamu-node](https://github.com/kamu-data/kamu-node): ❌
    DI update is needed and new configuration needs to be exposed - will be preparing a PR

### CODE REVIEW

File: src/app/cli/src/services/config/models.rs [hunk 1]
```diff
@@ -24,54 +25,59 @@ use url::Url;
 #[derive(Debug, Clone, Merge, Serialize, Deserialize)]
 #[serde(deny_unknown_fields, rename_all = "camelCase")]
 pub struct CLIConfig {
+    /// Database connection configuration
+    pub database: Option<DatabaseConfig>,
```
Comment (sergiimk) [inline src/app/cli/src/services/config/models.rs:29]: Sorry for the diff - sorted fields alphabetically

File: src/adapter/http/tests/tests/test_data_query.rs [hunk 2]
```diff
@@ -276,6 +285,208 @@ async fn test_data_query_handler_full() {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+#[test_group::group(engine, datafusion)]
+#[test_log::test(tokio::test)]
+async fn test_data_query_handler_v2() {
```
Comment (sergiimk) [inline src/adapter/http/tests/tests/test_data_query.rs:290]: This test explains it all

File: src/adapter/http/src/data/metadata_handler.rs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        .then(vis::SearchSetVocabVisitor::new);
+
+    let mut visitors: [&mut dyn MetadataChainVisitor<Error = vis::Infallible>; 6] = [
+        &mut attachments_visitor,
+        &mut info_visitor,
+        &mut license_visitor,
+        &mut schema_visitor,
+        &mut seed_visitor,
+        &mut vocab_visitor,
+    ];
```
Comment (sergiimk) [inline src/adapter/http/src/data/metadata_handler.rs:171]: FYI @s373r I'm super happy about this trick:
- Notice that all response elements are optional - so if user did not request `seed` or `schema` we ideally don't want to search for this event
- So I made all visitors above optional
- I then implemented `MetadataChainVisitor` for `Option<T> where T: MetadataChainVisitor`
- If the `Option` is `None` - the visitor will immediately `Stop`
- If `Some` - it delegates the calls to inner visitor

This is about as concise as it gets :)

Comment (s373r) [reply src/adapter/http/src/data/metadata_handler.rs:171]: @sergiimk , :100: 

Awesome idea with the optional visitors -- looking at the changes in the other PR noticed this, very cool!

**Pull request merged**